warehouse/schema/schema.go

```diff
@@ -78,56 +77,89 @@ type schema struct {
 	fetchSchemaRepo                  fetchSchemaRepo
 	now                              func() time.Time
 	cachedSchema                     model.Schema
-	cacheExpiry                      time.Time
 	cachedSchemaMu                   sync.RWMutex
 }
 
 func New(
-	db *sqlquerywrapper.DB,
+	ctx context.Context,
 	warehouse model.Warehouse,
 	conf *config.Config,
-	logger logger.Logger,
+	slogger logger.Logger,
 	statsFactory stats.Stats,
 	fetchSchemaRepo fetchSchemaRepo,
-) Handler {
+	schemaRepo schemaRepo,
+	stagingFileRepo stagingFileRepo,
+) (Handler, error) {
 	ttlInMinutes := conf.GetDurationVar(720, time.Minute, "Warehouse.schemaTTLInMinutes")
-	schema := &schema{
+	sh := &schema{
 		warehouse:                        warehouse,
-		log:                              logger.Child("schema"),
+		log:                              slogger.Child("schema"),
 		ttlInMinutes:                     ttlInMinutes,
-		schemaRepo:                       repo.NewWHSchemas(db),
+		schemaRepo:                       schemaRepo,
 		stagingFilesSchemaPaginationSize: conf.GetInt("Warehouse.stagingFilesSchemaPaginationSize", 100),
-		stagingFileRepo:                  repo.NewStagingFiles(db),
+		stagingFileRepo:                  stagingFileRepo,
 		fetchSchemaRepo:                  fetchSchemaRepo,
 		enableIDResolution:               conf.GetBool("Warehouse.enableIDResolution", false),
 		now:                              timeutil.Now,
 	}
-	schema.stats.schemaSize = statsFactory.NewTaggedStat("warehouse_schema_size", stats.HistogramType, stats.Tags{
+	sh.stats.schemaSize = statsFactory.NewTaggedStat("warehouse_schema_size", stats.HistogramType, stats.Tags{
 		"module":        "warehouse",
 		"workspaceId":   warehouse.WorkspaceID,
 		"destType":      warehouse.Destination.DestinationDefinition.Name,
 		"sourceId":      warehouse.Source.ID,
 		"destinationId": warehouse.Destination.ID,
 	})
-	return schema
+	// cachedSchema can be computed in the constructor
+	// we need not worry about it getting expired in the middle of the job
+	// since we need the schema to be the same for the entireduration of the job
+	whSchema, err := sh.schemaRepo.GetForNamespace(
+		ctx,
+		sh.warehouse.Source.ID,
+		sh.warehouse.Destination.ID,
+		sh.warehouse.Namespace,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("getting schema for namespace: %w", err)
+	}
+	if whSchema.Schema == nil {
+		sh.cachedSchema = model.Schema{}
+		return sh, nil
+	}
+	if whSchema.ExpiresAt.After(sh.now()) {
+		sh.cachedSchema = whSchema.Schema
+		return sh, nil
+	}
+	sh.log.Infon("Schema expired", obskit.DestinationID(sh.warehouse.Destination.ID), obskit.Namespace(sh.warehouse.Namespace), logger.NewTimeField("expiresAt", whSchema.ExpiresAt))
+	return sh, sh.fetchSchemaFromWarehouse(ctx)
```

Comment:
@shekhar-rudder 
`return sh, sh.fetchSchemaFromWarehouse(ctx)` : Don't this there should be any reason why it should not be fine to let constructor wait for time taking path for fetch schema as well. Just wanted to check once with you.

Reply:
> Don't think there should be any reason why it should not be fine to let constructor wait for time taking path for fetch schema as well

Didn't understand this. The constructor IS waiting for the schema fetch call to finish

Reply:
Ignore this. Resolving.